### ISSUE TITLE
Add a diagnostic for implicit relative imports

### ISSUE THREAD

**Issue by TyceHerrman** [astral-sh/ty#4096]
## Summary

Ty's module resolver has an internal fallback called "desperate resolution."
It can accept an unqualified import that resolves only by searching from the
importing file's directory. An import like that can work when the file is run
directly but then fail when it's imported as part of a package.

For example:

```text
project/
├── main.py
└── package/
    ├── __init__.py
    ├── foo.py
    └── bar.py
```

```python
# package/bar.py
import foo
```

```python
# main.py
import package.bar  # ModuleNotFoundError: No module named "foo"
```

Add a configurable diagnostic that:

- Reports an unqualified import when normal absolute resolution fails and the
  import succeeds only through this fallback from the importing file.
- Explains that the import may fail when the file is imported as a package.
- Suggests an explicit relative import such as `from . import foo` or the full
  absolute package path.
- Doesn't report imports resolved through configured roots, search paths, the
  standard library, or installed packages.

Basedpyright provides this behavior as
[`reportImplicitRelativeImport`](https://docs.basedpyright.com/latest/benefits-over-pyright/new-diagnostic-rules/#reportimplicitrelativeimport).

**Comment by carljm** [comment]
I can't reproduce the report given the exact tree as shown. We don't apply desperate resolution in that scenario, and we do error on `import foo`.

We do apply desperate resolution in cases where it's less clear where the intended import root is, due to a missing `__init__`.py. If `package/__init__.py` does not exist, then we aren't sure if `package` is a PEP 420 namespace package or just a directory with some scripts in it. In that case we err in the direction of leniency and allow the import.

I do think it would be reasonable to offer an opt-in diagnostic that fires when desperate module resolution is needed to resolve an import.

Thanks for the report!
